common-lib/src/cmr/common/generics.clj

```diff
@@ -117,3 +119,116 @@
+      (string/replace #"%uc-plural-generic%" (inf/plural (string/capitalize generic-type)))))
+
+;For each of the depth in the map get the output
+(defn build-markdown-toc
+  "Get the markdown file for a generic based on the api string being read in
+   Depth is a measure of how many spaces we need from the start of the line for the
+   markdown
+   Parameters:
+   * depth: integer a count of the number of spaces
+   * content: string, the content for the toc
+   * link: string, the link to the location in the document"
+  [depth content link]
+  (str (apply str (repeat depth " ")) (format "* [%s](#%s)\n" link content)))
+
+(defn get-toc-data
+  "Parses out the toc information from the api.md to build the markdown for the toc
+   Parameters:
+   * table-of-content-headers: string list, the headers in the document
+   * doc-type: string [ingest | search]"
+  ([table-of-content-headers doc-type]
+   (let [depth (if (= doc-type "ingest") (ingest-depth-map (count (re-seq #"#" table-of-content-headers))) (search-depth-map (count (re-seq #"#" table-of-content-headers))))
+         link (peek (re-find #"=\"(.*)\">" table-of-content-headers))
+         content (peek (re-find #">\s+([^<]*)$" table-of-content-headers))]
+     (build-markdown-toc depth link content))))
+
+(defn format-generic-toc
+  "Return the specific schema's documentation files given the schema keyword name and version number.
+   if the file cannot be read, return an empty string which will have no impact on the API document.
+   Parameters:
+   * file-name: [ingest | search]
+   * generic-keyword: [:grid | ...]
+   * generic-version: 0.0.1
+   Returns: string"
+  [file-name generic-keyword generic-version]
+  (try
+    (-> "schemas/%s/v%s/%s.md"
+        (format (name generic-keyword) generic-version (name file-name))
+        (io/resource)
+        (slurp)
+        (get-toc-headers-from-markdown)
+        (as-> xs (map #(get-toc-data % (name file-name)) xs))
+        (string/join)
+        (as-> xs (if (= file-name "ingest")
+        (str (fill-in-generic-name "* %uc-plural-generic%\n    * /providers/\\<provider-id>/%plural-generic%/\\<native-id>\n"
+        (name generic-keyword)) xs) (str "" xs))))
+    (catch Exception e (str ""))))
+
+(defn all-generic-docs-toc
+  "Parse over all of the generic documents and return their combined markdown as a string
+  Parameters:
+  * file-name: [ingest | search]
+  Returns: string"
+  [file-name]
+  (string/join (seq (for [[k,v] (latest-approved-documents)] (format-generic-toc file-name k (str v))))))
+
+(defn format-toc-into-doc
+  "To fit generic docs into the toc a few html tags must be removed
+   Parameters: the fully markdown for ingest or search toc
+   * generic-doc-toc: the full markdown data for the table of contents
+   for either ingest or search
+   Returns: string"
+  [generic-doc-toc]
+  (string/replace-first
+   (string/replace generic-doc-toc #"\s*<\/li>\n?<\/ul>$" "") #"<ul>\s*\n?\s*<li>" ""))
```

Comment:
should the second parameter to replace-first also be on it's own line?

Reply:
I don't think so I was trying to make it not go so far to the right but, I think it makes it harder to read this way, thank you

Reply:
actually the more I look at this, the more I think a -> would make this more clear

Reply:
Right something like this
`  (-> generic-doc-toc
      (string/replace  #"\s*<\/li>\n?<\/ul>$" "")
  (string/replace-first #"<ul>\s*\n?\s*<li>" ""))`